runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/graph/LengthPrefixUnknownCoders.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * <p>TODO: Support a dynamic list of well known coders using either registration or manual listing.
+ */
+public class LengthPrefixUnknownCoders {
+  private static final String BYTES_CODER_TYPE = "urn:beam:coders:bytes:0.1";
+  private static final String LENGTH_PREFIX_CODER_TYPE = "urn:beam:coders:length_prefix:0.1";
+  private static final Set<String> WELL_KNOWN_CODER_URNS =
+      ImmutableSet.of(
+          BYTES_CODER_TYPE,
+          "urn:beam:coders:kv:0.1",
+          "urn:beam:coders:varint:0.1",
+          "urn:beam:coders:interval_window:0.1",
+          "urn:beam:coders:stream:0.1",
+          LENGTH_PREFIX_CODER_TYPE,
+          "urn:beam:coders:global_window:0.1",
+          "urn:beam:coders:windowed_value:0.1");
+
+  /**
+   * Recursively traverse the coder tree and wrap the first unknown coder in every branch with a
+   * {@link LengthPrefixCoder} unless an ancestor coder is itself a {@link LengthPrefixCoder}. If
+   * {@code replaceWithByteArrayCoder} is set, then replace that unknown coder with a
+   * {@link ByteArrayCoder}.
+   *
+   * @param coderId The root coder contained within {@code coders} to start the recursive descent
+   * from.
+   * @param components Contains the root coder and all component coders.
+   * @param replaceWithByteArrayCoder whether to replace an unknown coder with a
+   * {@link ByteArrayCoder}.
+   * @return A {@link MessageWithComponents} with the
+   * {@link MessageWithComponents#getCoder() root coder} and its component coders.
+   */
+  public static RunnerApi.MessageWithComponents forCoder(
+      String coderId,
+      RunnerApi.Components components,
+      boolean replaceWithByteArrayCoder) {
+
+    MessageWithComponents.Builder builder = MessageWithComponents.newBuilder();
+    RunnerApi.Coder currentCoder = components.getCodersOrThrow(coderId);
+    if (LENGTH_PREFIX_CODER_TYPE.equals(currentCoder.getSpec().getSpec().getUrn())) {
+      if (replaceWithByteArrayCoder) {
+        String lengthPrefixComponentCoderId = generateUniqueId(
+            coderId + "-byte_array",
+            Sets.union(components.getCodersMap().keySet(),
+                builder.getComponents().getCodersMap().keySet()));
+        Coder.Builder byteArrayCoder = Coder.newBuilder();
+        byteArrayCoder.getSpecBuilder().getSpecBuilder().setUrn(BYTES_CODER_TYPE);
+        builder.getComponentsBuilder().putCoders(lengthPrefixComponentCoderId,
+            byteArrayCoder.build());
+        builder.getCoderBuilder()
+            .addComponentCoderIds(lengthPrefixComponentCoderId)
+            .getSpecBuilder()
+            .getSpecBuilder()
+            .setUrn(LENGTH_PREFIX_CODER_TYPE);
+      } else {
+        builder.setCoder(currentCoder);
+        builder.setComponents(components);
+      }
+    } else if (WELL_KNOWN_CODER_URNS.contains(currentCoder.getSpec().getSpec().getUrn())) {
+      RunnerApi.Coder.Builder updatedCoder = currentCoder.toBuilder();
```

Comment:
This seems sufficiently involved (given the recursive subcomponent rewriting) that it could stand being in a separate method. This might be true for the bulk of the `replaceWithByteArrayCoder` branches as well

Reply:
Split out to different methods covering the different cases.